versions/development/SPEC.md

```diff
@@ -1218,81 +1256,343 @@ $runtime = 'runtime' $ws* '{' ($ws* $runtime_kv $ws*)* '}'
+
+Container source locations should use the syntax defined by the individual container repository. For example an image defined as `ubuntu:latest` would conventionally refer a docker image living on `DockerHub`, while an image defined as `quay.io/bitnami/python` would refer to a `quay.io` repository. 
 
+The supported attribute types are:
+
+* `String`: A single container location
+* `Array[String]`: A list of container entries. **Note**: the ordering of the list does not imply any implicit preference or ordering of the containers. All images are expected to be the same, and therefore any choice would be equally valid.
+
+**Single Image**
 ```wdl
-task test {
-  input {
-    String ubuntu_version
+task single_image_test {
+  #....
+  runtime {
+    container: "ubuntu:latest"
   }
-  command {
-    python script.py
+```
+
+**Multiple Images**
+```wdl
+task multiple_image_test {
+  #.....
+  runtime {
+    container: ["ubuntu:latest", "gcr.io/standard-images/ubuntu:latest"]
   }
+}
+```
+
+
+#### cpu (**Optional**)
+
+The `cpu` key defines the _minimum_ CPU required for this task, which must be available prior to the engine starting execution. The engine does not need to provide the exact amount of CPU requested (depending on the underlying infrastructure restrictions), however it may ONLY provision more CPU than requested and not less. For example if the wdl requested `cpu: 0.5`, but only discrete values were supported, then the engine might choose to provision `1.0` cpu instead.  Values are expected to be a `Float` or an `Int`
+
+
+##### Default Value
+
+If the `cpu` key is not defined, then the engine must set a default value of `1` CPU. The task, therefore will be guaranteed to run with 1 CPU. The engine must not provide a mechanism to set their own default value for disks which circumvents this, ie `runtime-defaults`. 
+
+
+```wdl
+task cpu_example {
+  #....	
   runtime {
-    docker: "ubuntu:" + ubuntu_version
+    cpu: 8
   }
 }
+
 ```
 
-Most key/value pairs are arbitrary.  However, the following keys have recommended conventions:
+#### memory (**Optional**)
 
-#### docker
+The `memory` key defines the _minimum_ memory required for this task which must be available prior to the engine starting execution. The engine does not need to provide the exact amount of memory requested, however it may ONLY provision more memory then requested and not less. For example, if the wdl requested `1 GB` but only blocks of `4 GB` were available, the engine might choose to provision `4.0 GB` instead.  Two kinds of values are supported for this attribute:
```

Comment:
One other thought: by allowing engines to be *more* generous, do we risk running into the exact same portability problems we have now:

* I write a WDL without any `memory` specification.
* My WDL works on my engine, because my engine gives me `10GiB` of ram even though the "default" is only `2GiB`
* I publish my WDL and someone else tries to run it, but fails, because their engine respects the default `memory` amount of `2GiB` but my task really needs more than that?

Reply:
@cjllanwarne I struggled with this as well but I am not sure if there is any way around it. The nature of the beast is that different infrastructure will assign memory in different blocks. Ie Cromwell uses the google machine types which are all discrete values.

There must be a balance that we can strike here, and maybe that balance comes from moving the `maxMemory` into the `runtime` section. If the maxMemory is not set then the implication is the task will be assigend at least what is defined by the memory `memory` key. If the Execution engine does not support that specific memory amount, then it must assign the closest value?

Reply:
The other thing would potentially be to make the `maxMemory` a default of 2* the `memory` or some other feature like that

Reply:
Hmmm... it seems like the only way to **guarantee** portability is therefore to _require_ engines to retry failed tasks due to OOMs (potentially capped by a `maxMemory` hint if one is provided) - in other words make the scheme proposed in https://github.com/broadinstitute/cromwell/pull/5017 a requirement of the spec?

Reply:
I think I would favor an implicit maxMemory over required retries. Not every engine will know if a task failed because of an oom issue. Not to mention that approach violates the maxRetries policy,  which if set to 0 forbids retried of a task (for any reason).

Reply:
That sounds ok, I guess I don't mind what we call the idea of a maximum as long as it exists and is enforced. I think we're actually proposing almost exactly the same thing, but with different names:

* I would have `memory` be a maximum/"guaranteed success" value, and have a separate `try_this_memory_first` hint to tell the engine there might be a better/lower average value to try first. 
* You would have an implicit `maxMemory` value which an engine cannot go above, and use `memory` as a "don't go below this value" floor. But for an engine to be sure that it can run a WDL written against some other engine, it must eventually be willing to go up as high as `maxMemory` after a few tries, even if it's not its first preference.

The tie-in with `maxRetries` is interesting... Maybe the implied behavior is to divide the difference between min and max memory into an appropriate number of chunks so that the final retry happens to also use the `maxMemory` value.